### PR TITLE
frontend: Refactor symbol collecting

### PR DESCRIPTION
**Pull request by flofriday**
This only refactors the collection of symbols and distribution of symboltables, the actual resolving is left as future work.

### CODE REVIEW

File: vadl/test/vadl/ast/AsmDescriptionTests.java [hunk 1]
```diff
@@ -83,8 +83,8 @@ application binary interface ABI for ISA = {}
   void asmDescriptionDuplicateModifierString() {
     var prog = """
           instruction set architecture ISA = {
-            relocation HI ( symbol : Bits <32> ) -> Bits <16> = ( symbol >> 16 ) & 0xFFFF
-            relocation LO ( symbol : Bits <32> ) -> Bits <12> =   symbol         & 0xFFF
+            relocation HI ( symbol : Bits <32> ) -> Bits <16> = (( symbol >> 16 ) & 0xFFFF) as Bits<16>
+            relocation LO ( symbol : Bits <32> ) -> Bits <12> = (  symbol         & 0xFFF ) as Bits<16>
```
Comment (flofriday) [inline vadl/test/vadl/ast/AsmDescriptionTests.java:87]: @benjaminkasper99 I put the tests into the check command which also does typechecking and found this issue, I know it's not necessary for this test to be correctly typed but I think we should still aim for all tests to be correctly typed if that's not an aspect we want to test.

File: vadl/test/vadl/ast/TypecheckerProcessorTest.java [hunk 2]
```diff
@@ -41,13 +41,9 @@ void shouldThrow_duplicatedDefinitions() {
...
     var spec = base.formatted(body);
-    var ast = Assertions.assertDoesNotThrow(() -> VadlParser.parse(spec), "Cannot parse input");
-    var typechecker = new TypeChecker();
-    var throwable = assertThrows(DiagnosticList.class, () -> typechecker.verify(ast));
-    TestUtils.assertErrors(throwable,
-        "Contains multiple `firmware` definitions",
-        "Contains multiple `start` definitions"
+    var throwable = Assertions.assertThrows(DiagnosticList.class, () -> VadlParser.parse(spec)
     );
+    TestUtils.assertErrors(throwable, "Symbol name already used: start");
```
Comment (flofriday) [inline vadl/test/vadl/ast/TypecheckerProcessorTest.java:46]: @benjaminkasper99 The problem here is that this issue is now already found during symbol resolution instead of the typechecking.

I don't know if you still want to leave this test in but because the old symbol collector skipped some parts of the trees it previously was left as an task for the typechecker.

File: vadl/main/vadl/ast/SymbolTable.java [hunk 3]
```diff
@@ -380,489 +380,264 @@ private void reportAlreadyDefined(String error, SourceLocation location,
...
-        }
+    @Override
+    public void afterTravel(Definition definition) {
+      viamPath.pollLast();
+      symbolTables.pollLast();
+    }
+
+    @Override
+    public Void visit(AsmDescriptionDefinition definition) {
+      // More complex tasks like this require custom handling.
```
Comment (flofriday) [inline vadl/main/vadl/ast/SymbolTable.java:463]: **thought:** Only the more complex edge cases must be overriden and they all contain a comment on why it wasn't possible to use the default implementation

File: vadl/test/vadl/ast/AsmDescriptionTests.java (hunk 1 above)
Comment (benjaminkasper99) [reply vadl/test/vadl/ast/AsmDescriptionTests.java:87]: I agree, but I think the second cast here should then be `as Bits<12>`?

File: vadl/main/vadl/ast/SymbolTable.java (hunk 3 above)
Comment (benjaminkasper99) [reply vadl/main/vadl/ast/SymbolTable.java:463]: I think that's a fine solution. Also the adaptations for the `Asm...` definitions look good to me.

File: vadl/test/vadl/ast/AsmDescriptionTests.java (hunk 1 above)
Comment (flofriday) [reply vadl/test/vadl/ast/AsmDescriptionTests.java:87]: Yes you are right the second one should be 12 Bits

**Review by Jozott00 — APPROVED** [review]

**Review by benjaminkasper99 — APPROVED** [review]
